### PR TITLE
Move experimental frozen to frozen shard limit

### PR DESCRIPTION
**Pull request by henningandersen**
Frozen indices created on 7.12 would not belong to the frozen shard
limit group, now we convert them when last node is upgraded.

Relates #71392

~~This PR currently includes #71777, but that will be merged to master before this goes in.~~ (done)

Additionally, this PR has a few outstanding version related topics (todos), pending merge of #71760.

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java [hunk 1]
```diff
@@ -145,9 +147,17 @@ private static Settings buildIndexSettings(
...
             settings.put(SearchableSnapshotsConstants.SNAPSHOT_PARTIAL_SETTING.getKey(), true)
-                .put(DiskThresholdDecider.SETTING_IGNORE_DISK_WATERMARKS.getKey(), true)
-                .put(ShardLimitValidator.INDEX_SETTING_SHARD_LIMIT_GROUP.getKey(), ShardLimitValidator.FROZEN_GROUP);
+                .put(DiskThresholdDecider.SETTING_IGNORE_DISK_WATERMARKS.getKey(), true);
+
+            // we cannot apply this setting during rolling upgrade.
+            // todo: update version after backport
+            if (minNodeVersion.onOrAfter(Version.V_8_0_0)) {
+                settings.put(ShardLimitValidator.INDEX_SETTING_SHARD_LIMIT_GROUP.getKey(), ShardLimitValidator.FROZEN_GROUP);
+            }
```
Comment (henningandersen) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java:160]: All changes in this file are contained in #71777

**Comment by henningandersen** [comment]
Failure fixed by #71778

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/2

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java (hunk 1 above)
Comment (henningandersen) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: #71777 has now been merged and integrated here.

**Comment by henningandersen** [comment]
@elasticmachine update branch

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            .filter(
+                imd -> SearchableSnapshotsConstants.isPartialSearchableSnapshotIndex(imd.getSettings())
+                    && wrongShardLimitGroup(imd.getSettings())
+            )
+            .map(SearchableSnapshotIndexMetadataUpgrader::setShardLimitGroupFrozen)
+            .forEach(imd -> builder.put(imd, true));
+        return ClusterState.builder(currentState).metadata(builder).build();
+    }
+
+    private static boolean wrongShardLimitGroup(org.elasticsearch.common.settings.Settings settings) {
```
Comment (jasontedor) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java:106]: This is a nit: how about `notFrozenShardLimitGroup`?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            logger.info("Upgrading partial searchable snapshots to use frozen shard limit group");
+            clusterService.submitStateUpdateTask("searchable-snapshot-index-upgrader", new ClusterStateUpdateTask() {
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    return upgradeIndices(currentState);
+                }
+
+                @Override
+                public void onFailure(String source, Exception e) {
+                    logger.warn("upgrading frozen indices to have frozen shard limit group failed", e);
```
Comment (jasontedor) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java:73]: maybe add `, will retry on the next cluster state update`?

Comment (jasontedor) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java:73]: That is, we are letting the user know that something happened, at the warn level. Then they’re going to wonder what do they do? So let’s let them know, you don’t need to take any action, we’ll do it for you. Then, this is a purely informational error message from their perspective.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    private final ThreadPool threadPool;
+    private final AtomicBoolean upgraded = new AtomicBoolean();
+
+    public SearchableSnapshotIndexMetadataUpgrader(ClusterService clusterService, ThreadPool threadPool) {
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+    }
+
+    public void initialize() {
+        clusterService.addListener(this::clusterChanged);
```
Comment (jasontedor) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java:44]: Maybe we can remove this listener after the upgrade succeeds?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    private static IndexMetadata setShardLimitGroupFrozen(IndexMetadata indexMetadata) {
+        return IndexMetadata.builder(indexMetadata)
+            .settings(
+                Settings.builder()
+                    .put(indexMetadata.getSettings())
+                    .put(ShardLimitValidator.INDEX_SETTING_SHARD_LIMIT_GROUP.getKey(), ShardLimitValidator.FROZEN_GROUP)
+            )
+            .settingsVersion(indexMetadata.getSettingsVersion() + 1)
```
Comment (jasontedor) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java:117]: 😉

**Review by jasontedor — APPROVED** [review]
I left a few comments, but I don’t see a need for another round.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java (hunk 3 above)
Comment (henningandersen) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java]: ++, fixed in 01c1b3e2e93bebfb2468496f567cbb09be290be5

**Comment by henningandersen** [comment]
@elasticmachine update branch

**Comment by henningandersen** [comment]
@elasticmachine update branch

**Pull request merged**